Workflow Search refresh (#240)
temporalio/documentation: docs/tctl.md

[here](/docs/server-workflow-search/#memo-vs-search-attributes).

```

> tctl workflow start -tq hello-world -wt Workflow -et 60 -i '"temporal"' -memo_key ‘“Service” “Env” “Instance”’ -memo ‘“serverName1” “test” 5’

Reviewer: Maybe split this into multiline as well?

`--tls_server_name=<server name>` command line argument that passes an override value for the target server that is used for TLS host verification. It also enables host verification. The value must be one of the DNS names listed in the server TLS certificate.

TLS command line arguments can be provided via their respective environment variables to shorten the command line.

## Search Workflows

### List closed or open workflow executions

```
tctl workflow list
```

By default only one page is shown.
To view more items, use the `--more` flag.

```
tctl workflow list --more
```

Use the `--query` flag to list Workflows using an SQL-like query:

```
tctl workflow list \
  --query "WorkflowType='main.SampleParentWorkflow' AND CloseTime = missing"
```

This will return all open Workflows with `workflowType` as `main.SampleParentWorkflow`.

### List search attributes

You can query the list of search attributes with the following command:

> ```bash

Reviewer: You have bash on this but not the one above.

```bash
+---------------------+------------+
|         KEY         | VALUE TYPE |
+---------------------+------------+
| Status              | INT        |
| CloseTime           | INT        |
| CustomBoolField     | DOUBLE     |
| CustomDatetimeField | DATETIME   |
| CustomNamespace     | KEYWORD    |
| CustomDoubleField   | BOOL       |
| CustomIntField      | INT        |
| CustomKeywordField  | KEYWORD    |
| CustomStringField   | STRING     |
| NamespaceId         | KEYWORD    |
| ExecutionTime       | INT        |
| HistoryLength       | INT        |
| RunId               | KEYWORD    |
| StartTime           | INT        |
| WorkflowId          | KEYWORD    |
| WorkflowType        | KEYWORD    |
+---------------------+------------+
```

### Add new search attributes

Here is how you add a new search attribute:

```bash
tctl admin cluster add_search_attr \
  --search_attr_key <NewKey> \
  --search_attr_type string
```

The possible values for `--search_attr_type` are:

- string
- keyword
- int
- double
- bool
- datetime

### Start Workflow with Search Attributes

```bash
tctl workflow start \
  --taskqueue helloWorldGroup \
  --workflow_type main.Workflow \

>   --execution_timeout 60

Reviewer: Missing slash

`--tls_server_name=<server name>` command line argument that passes an override value for the target server that is used for TLS host verification. It also enables host verification. The value must be one of the DNS names listed in the server TLS certificate.

TLS command line arguments can be provided via their respective environment variables to shorten the command line.

## Search Workflows

### List closed or open workflow executions

```
./tctl workflow list

> # default will only show one page, to view more items, use --more flag

Reviewer 2: ```suggestion
# default will only show one page, to view more items, use `--more` flag
```

> ./tctl workflow list -m

Reviewer 2: ```suggestion
./tctl workflow list --more
```

```

> Use **--query** to list workflows with SQL like query:

Reviewer: I would love all of these to not require scroll bar.

Author: I did my best with this.
Each flag can be moved to a new line. But I am sure the command would break if the query itself is broken up, which is the longest part of the cmd. Might be able to do something creative - but then it would likely result in a decrease in user comprehension.

```

> ./tctl workflow list --query "WorkflowType='main.SampleParentWorkflow' AND CloseTime = missing "

Reviewer 2: ```suggestion
./tctl workflow list --query "WorkflowType='main.SampleParentWorkflow' AND CloseTime = missing"
```

```

> This will return all open workflows with workflowType as "main.SampleParentWorkflow".

Reviewer 2: ```suggestion
This will return all open workflows with `workflowType` as `main.SampleParentWorkflow`.
```

### Search attributes

You can query the list of search attributes with the following command:

```bash

> tctl --namespace samples-namespace cl get-search-attr

Reviewer 2: ```suggestion
tctl cluster get-search-attr
```

```

Here is some example output:

```bash
+---------------------+------------+
|         KEY         | VALUE TYPE |
+---------------------+------------+
| Status              | INT        |
| CloseTime           | INT        |
| CustomBoolField     | DOUBLE     |
| CustomDatetimeField | DATETIME   |
| CustomNamespace     | KEYWORD    |
| CustomDoubleField   | BOOL       |
| CustomIntField      | INT        |
| CustomKeywordField  | KEYWORD    |
| CustomStringField   | STRING     |
| NamespaceId         | KEYWORD    |
| ExecutionTime       | INT        |
| HistoryLength       | INT        |
| RunId               | KEYWORD    |
| StartTime           | INT        |
| WorkflowId          | KEYWORD    |
| WorkflowType        | KEYWORD    |
+---------------------+------------+
```

Here is how you add a new search attribute:

```bash

> tctl --namespace samples-namespace adm cl asa --search_attr_key NewKey --search_attr_type string

Reviewer 2: ```suggestion
tctl admin cluster add_search_attr --search_attr_key NewKey --search_attr_type string
```

```

The possible values for `--search_attr_type` are:

- string
- keyword
- int
- double
- bool
- datetime

### Start Workflow with Search Attributes

```bash

> tctl --ns samples-namespace workflow start --tq helloWorldGroup --wt main.Workflow --et 60 --dt 10 -i '"vancexu"' -search_attr_key 'CustomIntField | CustomKeywordField | CustomStringField |  CustomBoolField | CustomDatetimeField' -search_attr_value '5 | keyword1 | vancexu test | true | 2019-06-07T16:16:36-08:00'

Reviewer 2: ```suggestion
tctl workflow start --taskqueue helloWorldGroup --workflow_type main.Workflow --execution_timeout 60 --input '"john"' --search_attr_key 'CustomIntField | CustomKeywordField | CustomStringField |  CustomBoolField | CustomDatetimeField' --search_attr_value '5 | keyword1 | john test | true | 2019-06-07T16:16:36-08:00'
```

Reviewer 2: This is super ugly, btw. I need to change it.

```

### Search Workflows with List API

```bash

> tctl --ns samples-namespace wf list -q '(CustomKeywordField = "keyword1" and CustomIntField >= 5) or CustomKeywordField = "keyword2"' -psa

Reviewer 2: ```suggestion
tctl workflow list --query '(CustomKeywordField = "keyword1" and CustomIntField >= 5) or CustomKeywordField = "keyword2"' --print_search_attr
```

and all bellow
